File: app/src/main/java/com/example/uos_lms/core/domain/model/RepeatExam.java
package com.example.uos_lms.core.domain.model;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder(toBuilder = true)
@EqualsAndHashCode
@ToString
public class RepeatExam {
    private final String id;
    private final String examResultId;

    private final String studentUid;
    private final String studentName;
    private final String studentRollNumber;

    private final String subjectId;
    private final String subjectCode;
    private final String subjectTitle;

    private final double previousMarks;
    private final String previousGrade;
    private final double previousGpa;

    private final Double newMarks;
    private final String newGrade;
    private final Double newGpa;

    private final RepeatStatus repeatStatus;
    private final String rejectionReason;

    private final long createdAt;

    public boolean isEditable() {
        return repeatStatus == RepeatStatus.PENDING || repeatStatus == RepeatStatus.REJECTED;
    }
}

File: src/models/RepeatExam.js
const mongoose = require('mongoose');
const { RepeatStatus } = require('../constants/enums');

const { Schema } = mongoose;

// Extends a failed ExamResult rather than replacing it - previousMarks/Grade/Gpa keep the
// original attempt's history even after the repeat is approved and the ExamResult is
// updated in place (see repeatExamController.approve).
const repeatExamSchema = new Schema(
  {
    examResultId: { type: Schema.Types.ObjectId, ref: 'ExamResult', required: true },
    studentId: { type: String, ref: 'User', required: true },
    subjectId: { type: Schema.Types.ObjectId, ref: 'Subject', required: true },
    // Denormalized from the originating ExamResult at creation time, purely so the list
    // endpoint can scope an HOD to their own department without an extra join per row.
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', required: true },

    previousMarks: { type: Number, required: true },
    previousGrade: { type: String, required: true },
    previousGpa: { type: Number, required: true },

    newMarks: { type: Number, default: null },
    newGrade: { type: String, default: null },
    newGpa: { type: Number, default: null },

    repeatStatus: { type: String, enum: Object.values(RepeatStatus), default: RepeatStatus.PENDING },
    rejectionReason: { type: String, default: null },

    submittedBy: { type: String, ref: 'User', default: null },
    reviewedBy: { type: String, ref: 'User', default: null },
    reviewedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

module.exports = mongoose.model('RepeatExam', repeatExamSchema);
